### ISSUE TITLE
Allow underscore in domain name (which is perfectly legal)

### ISSUE THREAD

**Issue by cetex** [docker/docker#27447]
Trying to tag images to push to a certain docker registry i get:
docker tag <some image> docker-registry.service.dc_region.consul/ubuntu
Error parsing reference: "docker-registry.service.dc_region.consul/ubuntu" is not a valid repository/tag

However if i change the domain name to: docker-registry.service.dc**-**region.consul it works.

I've checked the tickets about hostnames already registered and i'm quite amazed that enforcing RFC1123 was considered a good / valid solution to a non-existing problem when the only result seems to be breaking working setups for people..

Docker version 1.12.1.

**Comment by troy0820** [comment]
I'm not sure but this document says that underscores are not allowed in the `docker tag`  command 
`The hostname must comply with standard DNS rules, but may not contain underscores`

https://github.com/docker/docker/blob/b248de7e332b6e67b08a8981f68060e6ae629ccf/man/docker-tag.1.md

**Comment by justincormack** [comment]
cc @stevvooe

**Comment by cetex** [comment]
I did a workaround by adding another CNAME pointing to the proper docker registry and retagging the images with the new one, quite annoying that i have to do this though when everything else (besides docker) seems to accept underscores in the domain name.

new dns record, this isn't a proper solution since we'll have to change it all the time but it worked for now:
docker-registry.service.consul IN CNAME docker-registry.service.dc_region.consul.

docker tag <image> docker-registry.service.consul/ubuntu
docker push docker-registry.service.consul/ubuntu

**Comment by stevvooe** [comment]
@cetex Could you provide a reference to an RFC that allow underscores for hostnames? I can't seem to find one. I think https://tools.ietf.org/html/rfc1034, https://tools.ietf.org/html/rfc1035 both only allow "-". https://tools.ietf.org/html/rfc5890 allows further extensions with punycode, and is closer to what docker allows but I can't see to find a complete grammar showing the allowance of underscores. As test, I even went over to hover.com to registry a few domain names with underscores and had no luck.

One has to remember that a docker image name _is not_ a hostname. It is a docker image name that has a number of other syntax restrictions, including the support of extracting a hostname from it. The grammar is a compromise between many of these constraints.

I'm sorry that this seems to have cause undue pain but I'm surprised to see that using underscores has worked as a way of naming a data center region in your infrastructure up until now. This is a common restriction and using cnames is the commonly suggested workaround.

Docker has used restricted hostname for image naming since its inception. We have opened it up slightly in the past few years to allow punycode and some other items, but, for the most part, it has been consistent.

**Comment by cetex** [comment]
I realize now that you're probably right!
"_" is allowed in dns, [rfc2782](http://ietf.org/rfc/rfc2782.txt) but not in use as part of hostnames, that's where i've mixed stuff up.

I get a bit annoyed when something is quick to enforce "proper conventions" where they're not required. I understand if this would be due to a hard requirement, but imho it doesn't seem like there's a hard requirement from your (dockers) perspective in this case.

For example, on docker 1.9.0 this actually works well:
`docker pull docker-registry.service.dc_region.consul/ubuntu`
And this:
`docker tag docker-registry.service.dc_region.consul/ubuntu docker-registry.service.dc_region.consul/ubuntu2`
This as well:
`docker pull docker-registry.service.dc_region.consul/ubuntu2`

```
docker images -a
REPOSITORY                                                           TAG                 IMAGE ID            CREATED             VIRTUAL SIZE
docker-registry.service.dc_region.consul/trusty/ubuntu    latest              31bc2fb07354        7 weeks ago         95.28 MB
docker-registry.service.dc_region.consul/trusty/ubuntu2   latest              31bc2fb07354        7 weeks ago         95.28 MB
```

On upgrade to 1.10.0 it seems like all containers with a non-rfc-approved hostname automatically has all "broken" tags removed, effectively making them useless and unusable:
time="2016-10-17T22:49:11.920322010+02:00" level=info msg="migrated layer 31bc2fb0768299a1bf6f5b25ae70857e4f1775d88fc15590f8e3feb1a6f3be4f to sha256:e8d8523be2f4b55fda356356a785f5bd6c90b5af72f1bf52ecadf1ccf81b2323"
time="2016-10-17T22:49:11.920833040+02:00" level=info msg="migrated image 31bc2fb0768299a1bf6f5b25ae70857e4f1775d88fc15590f8e3feb1a6f3be4f to sha256:0c57f55ddc3d313eb8334f963c1ad2174707d4eb54cbc45656ea87e113184c04"
time="2016-10-17T22:49:11.921163261+02:00" level=error msg="migrate tags: invalid name \"docker-registry.service.dc_region.consul/ubuntu\", \"invalid reference format\""
time="2016-10-17T22:49:11.921209202+02:00" level=error msg="migrate tags: invalid name \"docker-registry.service.dc_region.consul/ubuntu2\", \"invalid reference format\""
time="2016-10-17T22:49:11.921241342+02:00" level=info msg="Graph migration to content-addressability took 1.65 seconds"

```
docker images -a
REPOSITORY          TAG                 IMAGE ID            CREATED             SIZE
<none>              <none>              0c57f55ddc3d        7 weeks ago         95.28 MB
```

Also, none of the above commands with pull, tag or push works any more:
```
docker pull docker-registry.service.dc_region.consul/ubuntu
Error parsing reference: "docker-registry.service.dc_region.consul/ubuntu" is not a valid repository/tag
```

If it was an actual limitation of the DNS protocol itself I would understand the need to be strict about naming stuff, but the protocol doesn't seem to have this restriction implemented, so why should docker try to enforce it when it's just another user of the underlying DNS infrastructure?

If something should limit what hostnames are allowed it should be the dns system or the server (if it has anything to say about it's hostname, and if it even cares about it something which docker registry doesn't seem to do), the client should be quite forgiving in it's implementation, and all other clients do seem to be quite relaxed in how they work.

In short: It seems like it's the client (docker daemon) blocking an otherwise fully working setup due to adding a requirement to abide by an RFC that isn't required for technical reasons..

I didn't know that docker would strip tags from images if the naming scheme didn't follow the RFC on upgrades, and nothing of this is mentioned in the release notes for 1.10.0 as far as i can see.

/rant :)

**Comment by stevvooe** [comment]
@cetex Here is information about the migration in the [release notes](https://github.com/docker/docker/blob/master/CHANGELOG.md#distribution-6). Removing invalid references may not have been the right behavior but don't assume we made this change randomly or without technical requirements.

What would you like us to do?

**Comment by cetex** [comment]
I'd prefer it if you relaxed the hostname-part of the image naming a lot. :)
container hostnames could still be limited, if required for technical reasons, but i don't really see the need for it.
If pull, push and similar doesn't work because whatever name resolution in use can't resolve a host it's perfectly fine. And in this case it's not your problem anyways.

Take this fictional example if a new cool way to resolve hostnames that doesn't have the limitations of DNS surfaces:
Docker depends on the libc which in turn implements name resolution through [Name Service Switch](https://en.wikipedia.org/wiki/Name_Service_Switch)
As long as the hostnames aren't differing too much from the dns standard it could be a quite simple implementation to just add another module to nss which in turn resolved through this other mean of resolution.

Or just take this simple example, is this a valid hostname on linux?, it's at least accepted:
`echo "127.0.0.1 åäöa¤pa^¨~gnu_*.kurt" >> /etc/hosts`

```
# ping åäöa¤pa^¨~gnu_*.kurt -c1
PING åäöa¤pa^¨~gnu_*.kurt (127.0.0.1) 56(84) bytes of data.
64 bytes from localhost (127.0.0.1): icmp_seq=1 ttl=64 time=0.044 ms

--- åäöa¤pa^¨~gnu_*.kurt ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.044/0.044/0.044/0.000 ms
```

In our case we're still gonna have to rename stuff because we should conform somewhat to dns standards. (although *.consul being local to each datacenter doesn't work well nowadays when anyone can register a new GTLD, we can never be fully conforming to "public DNS standards" with this setup)
But renaming is a longer term project, it would be nice if docker accepted whatever naming scheme we use.

**Comment by thaJeztah** [comment]
I think this is a duplicate of https://github.com/docker/docker/issues/27275

**Comment by stevvooe** [comment]
@cetex Remember, that while images are partially a DNS name, they also are used as an image name. The fact that this worked pre-1.10 was an oversight -- image names weren't being correctly validated. The result was diverging support between satellite systems, manifesting as images that were un-pushable.

I don't think we can open up underscores in the hostname portion of image names.

**Comment by thaJeztah** [comment]
I agree with @stevvooe here, I don't think we should change this back to a faulty behavior.

**Comment by mostolog** [comment]
AFAIK, underscores aren't recommendable: http://stackoverflow.com/a/2183140
To sum up: 

- DNS spec allows underscores
- But underscores aren't valid in hostnames

Looking for an easier DNS-naming convention I won't accept underscores.

My 2 cents.

**Comment by cetex** [comment]
Ok, i give up. :)

It's not a hard requirement for our env to work anyways, but it would've been nice if services we run didn't try to enforce stuff where it's not required from the service's perspective.

**Comment by Anvil** [comment]
Sorry to bother you, but why is it working on ubuntu with official docker.com package, but not on fedora with fedora-rebuilt one ?

On Ubuntu:
```
$ docker pull os_registry.domain/asdf/centos:centos7 
centos7: Pulling from asdf/centos
a43f319e745a: Pull complete 
510abb6806d4: Pull complete 
ea20a52c0278: Pull complete 
8decb041f6a8: Pull complete 
Digest: sha256:116f86f957ec502f747456139a31c318694c7a0c69523526a6813ca6b866a745
Status: Downloaded newer image for os_registry.domain/asdf/centos:centos7

$ docker version
Client:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 21:44:32 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.3
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   6b644ec
 Built:        Wed Oct 26 21:44:32 2016
 OS/Arch:      linux/amd64

$ apt-cache show docker-engine
Package: docker-engine
Status: install ok installed
Priority: optional
Section: admin
Installed-Size: 99207
Maintainer: Docker <support@docker.com>
Architecture: amd64
Version: 1.12.3-0~trusty
Depends: iptables, init-system-helpers (>= 1.13~), sysv-rc (>= 2.88dsf-24) | file-rc (>= 0.8.16), libapparmor1 (>= 2.6~devel), libc6 (>= 2.17), libdevmapper1.02.1 (>= 2:1.02.63), libltdl7 (>= 2.4.2), libsystemd-journal0 (>= 201)
Recommends: aufs-tools, ca-certificates, cgroupfs-mount | cgroup-lite, git, xz-utils, apparmor
Conflicts: docker (<< 1.5~), docker-engine-cs, docker.io, lxc-docker, lxc-docker-virtual-package
Conffiles:
 /etc/init.d/docker 29701769de7c7f966d739905f120ed57
 /etc/default/docker ddb6348855959363f63b73853c21f4d0
 /etc/init/docker.conf a3cf502224ffae47dd79f59c82cdd332
 /etc/bash_completion.d/docker 68cb0205028874522170850d8849d7ca
Description: Docker: the open-source application container engine
 Docker is an open source project to build, ship and run any application as a
 lightweight container
 .
 Docker containers are both hardware-agnostic and platform-agnostic. This means
 they can run anywhere, from your laptop to the largest EC2 compute instance and
 everything in between - and they don't require you to use a particular
 language, framework or packaging system. That makes them great building blocks
 for deploying and scaling web apps, databases, and backend services without
 depending on a particular stack or provider.
Description-md5: 081115dab45344eb40852d38ab20e419
Homepage: https://dockerproject.org
```

On fedora 25:
```
$ docker pull os_registry.domain:443/asdf/centos:centos7
Error parsing reference: "os_registry.domain:443/asdf/centos:centos7" is not a valid repository/tag
$ docker pull os_registry.domain/asdf/centos:centos7
Trying to pull repository docker.io/os_registry.domain/asdf/centos ... 
^Czsh: exit 130

$ docker version   
Client:
 Version:         1.12.6
 API version:     1.24
 Package version: docker-common-1.12.6-5.git037a2f5.fc25.x86_64
 Go version:      go1.7.4
 Git commit:      037a2f5/1.12.6
 Built:           Wed Jan 18 12:11:29 2017
 OS/Arch:         linux/amd64

Server:
 Version:         1.12.6
 API version:     1.24
 Package version: docker-common-1.12.6-5.git037a2f5.fc25.x86_64
 Go version:      go1.7.4
 Git commit:      037a2f5/1.12.6
 Built:           Wed Jan 18 12:11:29 2017
 OS/Arch:         linux/amd64

$ rpm -qi docker-common
Name        : docker-common
Epoch       : 2
Version     : 1.12.6
Release     : 5.git037a2f5.fc25
Architecture: x86_64
Install Date: Wed 18 Jan 2017 04:41:34 PM CET
Group       : Unspecified
Size        : 2483
License     : ASL 2.0
Signature   : (none)
Source RPM  : docker-1.12.6-5.git037a2f5.fc25.src.rpm
Build Date  : Wed 18 Jan 2017 01:14:20 PM CET
Build Host  : buildvm-25.phx2.fedoraproject.org
Relocations : (not relocatable)
Packager    : Fedora Project
Vendor      : Fedora Project
URL         : https://github.com/projectatomic/docker
Summary     : Common files for docker and docker-latest
Description :
This package contains the common files /usr/bin/docker which will point to
/usr/bin/docker-current or /usr/bin/docker-latest configurable via
/etc/sysconfig/docker
```

My coworker running arch tells it does work on his system...

Please, what should I trust ? What can I expect in near future ?

**Comment by Anvil** [comment]
Self reply: I understand the `URL : https://github.com/projectatomic/docker` in the `rpm -qi` output and the consequences. But I feel like docker is not respecting its own standard here.

**Comment by thaJeztah** [comment]
@Anvil are you having the same issue with docker 1.12.6 on Ubuntu? The Red Hat fork of Docker has so many changes applied that it's not really worth comparing it.

**Comment by Anvil** [comment]
@thaJeztah: as far as i understand the situation, official docker 1.12.6 accepts underscores on all platforms, despite of the fact that the `man docker-tag` saying that 
> The hostname must comply with standard DNS rules, but may not contain underscores.

Red Hat's fork does not allow underscores in the tag hostname part, but I understand :) you're not responsible for it.

So my question is: should the docker documentation be updated to reflect reality or is this a bug in the code being too permissive when it should not ?

**Comment by mostolog** [comment]
@Anvil @thaJeztah As commented before, according to http://stackoverflow.com/a/2183140:

- DNS spec allows underscores
- Underscores aren't valid in hostnames

I suggest docker team to consider a clear statement on this, involving https://github.com/docker/swarmkit/issues/1242: Might docker allow underscores but emit a warning if there are any?

**Comment by thaJeztah** [comment]
@mostolog @Anvil I think that discussion has to be done on the https://github.com/docker/distribution repository, as that's where the specs for these are defined; https://github.com/docker/distribution/blob/11cb04e994033c5ec98e6630f6e0c6a2b2677344/reference/reference.go#L8-L10

**Comment by Anvil** [comment]
@thaJeztah Thanks for the pointers. I'll follow @mostolog newly opened entry with much interest. But in the meantime here is what I understand of the current situation.

https://github.com/docker/distribution/blob/11cb04e994033c5ec98e6630f6e0c6a2b2677344/reference/regexp.go#L25 clearly shows that underscores is not allowed. I think it's safe to say that docker/distribution respects its own standard. 🥇 

I have to confess we are currently using an artifactory registry, at my place. And I think this is the reason why it has even worked in the first place.

In the end, I think docker/docker is too permissive. Maybe this should be considered a bug if things will not evolve on the docker/distribution side.

**Comment by NamPNQ** [comment]
I think if not allow underscore in the domain name, should not allow it in `docker login` too
